Source/WebKit/WebProcess/WebPage/FindController.cpp

```diff
@@ -136,38 +136,35 @@ static RefPtr<LocalFrame> frameWithSelection(Page* page)
     return nullptr;
 }
 
-void FindController::updateFindUIAfterPageScroll(bool found, const String& string, OptionSet<FindOptions> options, unsigned maxMatchCount, DidWrap didWrap, FindUIOriginator originator)
+void FindController::updateFindUIAfterPageScroll(bool found, const String& string, OptionSet<FindOptions> options, unsigned maxMatchCount, DidWrap didWrap, FindUIOriginator originator, std::optional<FrameIdentifier> idOfFrameContainingString, CompletionHandler<void(std::optional<WebCore::FrameIdentifier>, Vector<IntRect>, uint32_t, int32_t, bool)>&& completionHandler)
 {
-    if (options.contains(FindOptions::DoNotSetSelection))
-        return;
-
     RefPtr selectedFrame = frameWithSelection(m_webPage->corePage());
 
 #if ENABLE(PDF_PLUGIN)
     auto* pluginView = mainFramePlugIn();
 #endif
 
     bool shouldShowOverlay = false;
-
+    unsigned matchCount = 0;
     if (!found) {
 #if ENABLE(PDF_PLUGIN)
         if (!pluginView)
 #endif
             m_webPage->corePage()->unmarkAllTextMatches();
 
-        if (selectedFrame)
+        if (selectedFrame && !options.contains(FindOptions::DoNotSetSelection))
             selectedFrame->selection().clear();
 
         hideFindIndicator();
         resetMatchIndex();
         didFailToFindString();
-
-        m_webPage->send(Messages::WebPageProxy::DidFailToFindString(string));
+        if (originator != FindUIOriginator::FindString)
+            m_webPage->send(Messages::WebPageProxy::DidFailToFindString(string));
```

Comment:
Can't we make this a completionHandler(std::nullopt, ...) or something?  It isn't great for there to be a completion handler that takes a result and another conditional message that also sends the results.

Reply:
This PR makes changes so `FindString()` uses the completion handler instead of `DidFindString()` and `DidFailToFindString()`. There is also `FindStringMatches()` (for some reason) that still relies on this IPC message being sent.

Reply:
There would need to be more changes to make `FindStringMatches()` use a completion handler. I'm fine with doing that in this PR, but it seemed like maybe I shouldn't?

Reply:
Separate PR is fine but it really should be done.  This code is crazy without that.